### ISSUE TITLE
More workload-driven benchmarks

### ISSUE THREAD

**Issue by caio** [cep21/tdigestbench#1]
Hey, really cool stuff! Thanks a lot for doing it.

Maintainer of the `caio/go-tdigest` package here and I'm pretty happy to see that my expectations of how it behaves hold true under external scrutiny :)

I'm not familiar with the influxdata implementation, but from your analysis I'm guessing they buffer the adds, which is probably what explains its great performance on `Add` and the slightly higher memory usage.

Originally I made the Add on the package I maintain very fast by using a fenwick tree to cache the hot part of the code, but [more realistic benchmarks culminated in its removal][1] by one of the many great contributors.

[1]: https://github.com/caio/go-tdigest/pull/25

You might be interested in cherry-picking those for your performance analysis. My expectation is that go-tdigest would perform competitively albeit consistently slower than a buffered implementation.

**Comment by caio** [comment]
[It is buffered indeed][1] so it probably also performs a *lot* better than go-tdigest against adversarial input (sorted samples, for example). Pretty cool!

[1]: https://github.com/influxdata/tdigest/blob/master/tdigest.go#L71

**Comment by cep21** [comment]
Thanks!  The pr was a bit hard to parse.  Which specific benchmark do you recommend i add?

**Comment by caio** [comment]
Yeah, I should have been clearer. It makes sense to look at individual `Add` calls, but it also makes sense to look at the performance of adding *all* samples. That's because of how the data structure behaves: an Add might be a simple increment; Might create a new centroid; And might merge centroids - All of these can be implemented differently and depend on the previous state of the digest.

In go-tdigest's case, just benchmarking `Add` wouldn't properly verify the impact of the internal implementation (where a cache is rebuilt based on the input and current state) because the slow case doesn't happen enough times, it just ends up being discarded as outlier.

Eh, still confusing I think. What I'm trying to say is that given how digests are used (add a bunch of samples, then do something with them), the outliers *do* matter, so looking at the performance of "adding a bunch of sample" is a better way to reason about its performance. (Edit: Which is the difference between BenchmarkTDigestAddOnce and BenchmarkTDigestAddMulti in the patch)

**Comment by cep21** [comment]
My add benchmark isn't a single add to a new digest each time, it's b.N adds to the same digest.

```
	for i := 0; i < b.N; i++ {
		tdigest.Add(source.Float64())
	}
```

Go's benchmark output reports the number of iterations used.  For caio, it was `1000000`

**Comment by caio** [comment]
Yup, that's what I mean. Your benchmark is the same as [BenchmarkTDigestAddOnce][1] and I'm saying that it also makes sense to have something that's more like [BenchmarkTDigestAddMulti][2], because there might be very high variance on `Add`.

[1]: https://github.com/caio/go-tdigest/blob/master/tdigest_test.go#L676
[2]: https://github.com/caio/go-tdigest/blob/master/tdigest_test.go#L704

**Comment by caio** [comment]
If it's still unclear what I mean, feel free to close and I'll come back with a PR with what I mean, but I can't promise when I'll be able to do it anytime soon. Keep up the good work!

**Comment by cep21** [comment]
So you're making the same data series of 10,000 numbers, then adding that same data series again and again to evaluate the performance of Add?  Won't this variance get evened out if I add 1,000,000 numbers (like I do in my benchmark).

**Comment by caio** [comment]
In most cases yes, but when there are rare occurrences of heavy outliers things get weird.

Perhaps an absurd example as code explains what I mean better:

hue_test.go:
```go
package hue

import (
	"testing"
	"time"
)

type Counter struct {
	tally uint64
}

func New() Counter {
	return Counter{tally: 0}
}

func (c *Counter) Inc() {
	c.tally++
}

func (c *Counter) SpikyInc() {
	c.tally++
	if c.tally%alot == 0 {
		time.Sleep(10 * time.Second)
	}
}

func BenchmarkInc(b *testing.B) {
	a := New()
	for n := 0; n < b.N; n++ {
		a.Inc()
	}
}

func BenchmarkSpikyInc(b *testing.B) {
	a := New()
	for n := 0; n < b.N; n++ {
		a.SpikyInc()
	}
}

const alot = 1000000000

func BenchmarkIncALot(b *testing.B) {
	a := New()
	for n := 0; n < b.N; n++ {
		for n := 0; n < alot; n++ {
			a.Inc()
		}
	}
}

func BenchmarkSpikyIncALot(b *testing.B) {
	a := New()
	for n := 0; n < b.N; n++ {
		for n := 0; n < alot; n++ {
			a.SpikyInc()
		}
	}
}
```

Which gives:

```
$ go test -bench . 
goos: linux
goarch: amd64
BenchmarkInc-4                  498365247                2.36 ns/op
BenchmarkSpikyInc-4             502002384                2.36 ns/op
BenchmarkIncALot-4                     1        2843448153 ns/op
BenchmarkSpikyIncALot-4                1        12397430095 ns/op
PASS
ok      _/home/caio/download/hue        18.091s
```

Notice how the individual Inc() vs SpikyInc() fails to capture the 10s spike.

**Comment by cep21** [comment]
I see

Your number of iterations caps at `502_002_384` which isn't big enough to hit `alot = 1_000_000_000` before Go stops your benchmark (since by default they try to take 1sec).  My computer is fast enough to reach the 1sec threshold so I get

```
BenchmarkInc-8                  935068092                1.28 ns/op
BenchmarkSpikyInc-8             1000000000              11.2 ns/op
```

I think you're seeing this because you let the test for `IncALot` run longer than the test for `Inc`.  Running longer allows it to reach `const alot` iterations.  Because the test runs longer, it gets more accurate results.  Try running the same test, but with a different flag.

`go test -bench . -benchtime 10s .`


Also, don't try to increment `const alot` because Go's benchmarks don't allow tests to run with more than 1e9 iterations.

**Comment by caio** [comment]
Alright, I'm failing to explain what I mean :sweat_smile: . Assuming I'm actually making sense, I'll come back with more solid stuff if/when time permits.
